src/draco/c_api/c_api.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 The Draco Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef DRACO_C_API_H_
+#define DRACO_C_API_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// If compiling with Visual Studio.
+#if defined(_MSC_VER)
+#define EXPORT_API __declspec(dllexport)
+#else
+// Other platforms don't need this.
```

Comment:
GCC and Clang do need an EXPORT_API definition since GCC v4
It is strongly recommended not to export all functions, especially for a C-API as it greatly reduces the chance of name collisions and speeds up the dylib/so loader.

The draco C-API so/dylib should be compiled with the flag -fvisibility=hidden, and the following added for GCC and clang compilers:
```
#ifdef DRACO_BUILDING_DLL // also update draco_targets.cmake
#define EXPORT_API __attribute__ ((visibility ("default")))
#else
#define EXPORT_API
#endif
```

Reply:
This is a little bit tricky to implement with the current approach of building the C-API as part of the "core" library, but I do agree it should happen.

The problem I'm facing if only the C-API is exported then the executables (`draco_decoder` and `draco_encoder`) nor the test can be correctly linked as those depend on the C++ API.

I'm starting to lean towards having a separate library for the C-API, as the CMake orchestration would be much more easy to manage and the `DRACO_C_API` will be orthogonal to other features. What do you think @RichardTea?

Reply:
I think you're right. While I'm sure it _possible_ to build all four, it'd be quite complex.

I think it's reasonable to pick an API for your shared libraries at configure time, as I'm reasonably sure that most consumers of the C-API will _not_ want the C++ version of the dynamic library, so no need to jump through the hoops to make them both happen.

I _think_ draco_encoder, draco_decoder and the C-API shared lib could all consume the same C++ static library "privately" to achieve the goal of hiding all the internal symbols - IIRC, symbols aren't ever truly private in a static library so that should still work.

Reply:
I've finally moved the C-API to its own library called `cdraco.[dll/lib/so/a/...]`, gated by the `DRACO_C_API` flag. This library is self contained and the dynamic version just exports the C-API, all the C++ stuff is hidden.